src/ApiService/ApiService/Log.cs

```diff
@@ -1,418 +1,435 @@
+                }
             }
         }
-        return new LogTracer(CorrelationId, newTags, _loggers, _logSeverityLevel);
+        PopulateTags(telemetryItem);
     }
 
-    public void Verbose(LogStringHandler message) {
-        if (_logSeverityLevel <= SeverityLevel.Verbose) {
-            var caller = GetCaller();
-            foreach (var logger in _loggers) {
-                logger.Log(CorrelationId, message, SeverityLevel.Verbose, Tags, caller);
-            }
-        }
-    }
+    /// <summary>
+    /// 
+    /// </summary>
+    /// <param name="telemetryItem"></param>
+    private static void PopulateTags(ISupportProperties telemetryItem) {
+        IDictionary<string, string> dict = telemetryItem.Properties;
 
-    public void Info(LogStringHandler message) {
-        if (_logSeverityLevel <= SeverityLevel.Information) {
-            var caller = GetCaller();
-            foreach (var logger in _loggers) {
-                logger.Log(CorrelationId, message, SeverityLevel.Information, Tags, caller);
+        var ourActivities = new LinkedList<Activity>();
+        {
+            var activity = Activity;
+            while (activity is not null && string.Equals(activity.OperationName, TagsActivityName)) {
+                _ = ourActivities.AddFirst(activity);
+                activity = activity.Parent;
             }
         }
-    }
 
-    public void Warning(LogStringHandler message) {
-        if (_logSeverityLevel <= SeverityLevel.Warning) {
-            var caller = GetCaller();
-            foreach (var logger in _loggers) {
-                logger.Log(CorrelationId, message, SeverityLevel.Warning, Tags, caller);
+        foreach (var activity in ourActivities) {
+            foreach (var kv in activity.Tags) {
+                dict[kv.Key] = Convert.ToString(kv.Value, CultureInfo.InvariantCulture) ?? $"Failed to convert {kv.Value}";
             }
         }
     }
+}
 
-    public void Error(LogStringHandler message) {
-        if (_logSeverityLevel <= SeverityLevel.Error) {
-            var caller = GetCaller();
-            foreach (var logger in _loggers) {
-                logger.Log(CorrelationId, message, SeverityLevel.Error, Tags, caller);
-            }
-        }
-    }
 
-    public void Critical(LogStringHandler message) {
-        if (_logSeverityLevel <= SeverityLevel.Critical) {
-            var caller = GetCaller();
-            foreach (var logger in _loggers) {
-                logger.Log(CorrelationId, message, SeverityLevel.Critical, Tags, caller);
+/// <summary>
+/// 
+/// </summary>
+public static class OneFuzzLoggerExt {
+    private static EventId EmptyEventId = new EventId(0);
+
+    /// <summary>
+    /// 
+    /// </summary>
+    /// <returns></returns>
+    public static string? GetCorrelationId(this ILogger _) {
+        foreach (var tag in OneFuzzLogger.Activity.Tags) {
```

Comment:
TryGetValue?

Reply:
Tags are IEnumerable  of KeyValue pairs,.

It is allowed to add multiple values with the same key, and they will all add all of them.

There is no method TryGetValue at all in the Tags collection